src/elgamal.rs

```diff
@@ -0,0 +1,244 @@
+use crate::utils;
+use encoding::all::UTF_16LE;
+use encoding::{DecoderTrap, EncoderTrap, Encoding};
+use rug::ops::Pow;
+use rug::{rand::RandState, Complete, Integer};
+use codec::{Decode, Encode};
+
+use crate::generic::{PublicKey,PrivateKey,Encryption};
+use sp_core::U256;
+use crate::utils::{bigint_u256, u256_bigint};
+
+const STR_RADIX: i32 = 10;
+const SEARCH_LIMIT: u32 = 100;
+
+/// Generate a seed data slice from a key data.
+pub trait Seed {
+    fn yield_seed(self) -> Integer;
+}
+
+impl Seed for PublicKey{
+    fn yield_seed(self) -> Integer {
+        let sum = self.p + self.g + self.h;
+        sum.to_owned()
+    }
+}
+
+/// The raw public key type use bytes string.
+#[derive(Clone, PartialEq, Eq, Encode, Decode, Debug)]
+pub struct RawPublicKey {
+    pub p: U256,
+    pub g: U256,
+    pub h: U256,
+    pub bit_length: u32,
+}
+
+/// To and from raw bytes of a public key. Use little endian byte order by default.
+pub trait RawKey {
+    fn to_raw(self) -> RawPublicKey;
+    fn from_raw(raw_key: RawPublicKey) -> Self;
+}
+
+impl RawKey for PublicKey {
+    fn to_raw(self) -> RawPublicKey {
+        RawPublicKey {
+            p: bigint_u256(&self.p),
+            g: bigint_u256(&self.g),
+            h: bigint_u256(&self.h),
+            bit_length: self.bit_length,
+        }
+    }
+
+    fn from_raw(raw_key: RawPublicKey) -> Self {
+        PublicKey{
+            p:u256_bigint(&raw_key.p),
+            g:u256_bigint(&raw_key.g),
+            h:u256_bigint(&raw_key.h),
+            bit_length: raw_key.bit_length,
+        }
+    }
+}
+
+
+///generate public_key with seed、bit_length、i_confidence
+///Generates public key K1 (p, g, h) and private key K2 (p, g, x).
+pub fn generate_pub_key(rand: &mut RandState, bit_length: u32, seed:Integer) -> PublicKey {
+    rand.seed(&seed);
+    let p = utils::random_prime_bigint(rand, bit_length.clone());
+    rand.seed(&seed);
```

Comment:
I don't think we still need to reset the seed here since the original step was designed to replicate python results. Since the mismatch of PRNG algorithm, as long as the generation is deterministic, it shall work OK. So I would like to remove those resetting steps to get better performance.